src/nbl/video/CVulkanPhysicalDevice.h

```diff
@@ -101,72 +100,163 @@ class CVulkanPhysicalDevice final : public IPhysicalDevice
+    // Todo(achal)
+    IDebugCallback* getDebugCallback() override { return nullptr; }
+
+    void getAvailableFormatsForSurface(const ISurface* surface, uint32_t& formatCount,
+        ISurface::SFormat* formats) const override
     {
-        const ISurfaceVK* vk_surface = static_cast<const ISurfaceVK*>(surface); // Todo(achal): This is problematic, if passed `surface` isn't a vulkan surface
+        constexpr uint32_t MAX_SURFACE_FORMAT_COUNT = 1000u;
+
+        if (surface->getAPIType() != EAT_VULKAN)
+            return;
 
-        uint32_t formatCount = 0u;
-        vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkphysdev, vk_surface->m_surface, &formatCount, nullptr);
-        std::vector<VkSurfaceFormatKHR> formats(formatCount);
-        vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkphysdev, vk_surface->m_surface, &formatCount, formats.data());
+        // Todo(achal): not sure yet, how would I handle multiple platforms without making
+        // this function templated
+        VkSurfaceKHR vk_surface = static_cast<const CSurfaceVulkanWin32*>(surface)->getInternalObject();
 
-        std::vector<ISurface::SFormat> result(formatCount);
+        VkResult retval = vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkPhysicalDevice, vk_surface,
+            &formatCount, nullptr);
+
+        // Todo(achal): Would there be a need to handle VK_INCOMPLETE separately?
+        if ((retval != VK_SUCCESS) && (retval != VK_INCOMPLETE))
+        {
+            formatCount = 0u;
+            return;
+        }
+
+        if (!formats)
+            return;
+
+        VkSurfaceFormatKHR vk_formats[MAX_SURFACE_FORMAT_COUNT];
+        retval = vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkPhysicalDevice, vk_surface,
+            &formatCount, vk_formats);
+
+        // Todo(achal): Would there be a need to handle VK_INCOMPLETE separately?
+        if ((retval != VK_SUCCESS) && (retval != VK_INCOMPLETE))
+        {
+            formatCount = 0u;
+            formats = nullptr;
+            return;
+        }
 
         for (uint32_t i = 0u; i < formatCount; ++i)
         {
-            result[i].format = ISurfaceVK::getFormat(formats[i].format);
-            result[i].colorSpace = ISurfaceVK::getColorSpace(formats[i].colorSpace);
+            formats[i].format = getFormatFromVkFormat(vk_formats[i].format);
+            formats[i].colorSpace = getColorSpaceFromVkColorSpaceKHR(vk_formats[i].colorSpace);
         }
+    }
+    
+    ISurface::E_PRESENT_MODE getAvailablePresentModesForSurface(const ISurface* surface) const override
+    {
+        constexpr uint32_t MAX_PRESENT_MODE_COUNT = 4u;
+
+        if (surface->getAPIType() != EAT_VULKAN)
+            return ISurface::EPM_UNKNOWN;
+
+        // Todo(achal): not sure yet, how would I handle multiple platforms without making
+        // this function templated
+        VkSurfaceKHR vk_surface = static_cast<const CSurfaceVulkanWin32*>(surface)->getInternalObject();
```

Comment:
again, self reporting would be best

basically `ISurface` needs virtual methods for both format and presentation mode listing.

Reply:
and surface caps

Reply:
resolved